schemas/iso19139/src/main/plugin/iso19139/process/onlinesrc-add.xsl

```diff
@@ -135,14 +130,72 @@ Insert is made in first transferOptions found.
+    <xsl:copy>
+      <xsl:apply-templates select="node()|@*"/>
+      <gmd:MD_Distribution>
+        <gmd:transferOptions>
+          <gmd:MD_DigitalTransferOptions>
+            <xsl:call-template name="createOnlineSrc"/>
+          </gmd:MD_DigitalTransferOptions>
+        </gmd:transferOptions>
+      </gmd:MD_Distribution>
+    </xsl:copy>
+  </xsl:template>
+
+  <!--  <gmd:transferOptions> does not exist-->
+  <xsl:template match="*/gmd:distributionInfo[1]/gmd:MD_Distribution[not(gmd:transferOptions) and $update_flag = false() and position() = 1]">
+    <xsl:copy>
+      <xsl:apply-templates select="node()|@*"/>
+      <gmd:transferOptions>
+        <gmd:MD_DigitalTransferOptions>
+          <xsl:call-template name="createOnlineSrc"/>
+        </gmd:MD_DigitalTransferOptions>
+      </gmd:transferOptions>
+    </xsl:copy>
+  </xsl:template>
+
+  <!--  <gmd:MD_DigitalTransferOptions> does not exist-->
+  <xsl:template match="*/gmd:distributionInfo[1]/gmd:MD_Distribution[1]/gmd:transferOptions[not(gmd:MD_DigitalTransferOptions) and $update_flag = false() and position() = 1]">
+    <xsl:copy>
+      <xsl:apply-templates select="node()|@*"/>
+      <gmd:MD_DigitalTransferOptions>
+        <xsl:call-template name="createOnlineSrc"/>
+      </gmd:MD_DigitalTransferOptions>
+    </xsl:copy>
+  </xsl:template>
+
+  <!--  Add new gmd:gmd:onLine-->
+  <xsl:template match="*/gmd:distributionInfo[1]/gmd:MD_Distribution[1]/gmd:transferOptions[1]/gmd:MD_DigitalTransferOptions[$update_flag = false() and position() = 1]">
+    <xsl:copy>
+      <xsl:apply-templates select="@*"/>
+      <xsl:apply-templates
+        select="gmd:unitsOfDistribution|
+                gmd:transferSize|
+                gmd:onLine"/>
+
+      <xsl:call-template name="createOnlineSrc"/>
+
+      <xsl:apply-templates select="gmd:offLine"/>
+    </xsl:copy>
+  </xsl:template>
 
-  <!-- Updating the link matching the update key. -->
-  <xsl:template match="gmd:onLine[$updateKey != '' and
-                        normalize-space($updateKey) = concat(
-                        gmd:CI_OnlineResource/gmd:linkage/gmd:URL,
-                        gmd:CI_OnlineResource/gmd:protocol/*,
-                        gmd:CI_OnlineResource/gmd:name/gco:CharacterString)
-                        ]">
+  <!-- End of inserting gmd:onLine -->
+
+
+  <!-- Updating the gmd:onLine based on update parameters -->
+  <!-- Note: first part of the match needs to match the xsl:for-each select from extract-relations.xsl in order to get the position() to match -->
+  <!-- The unique identifier is marked with resourceIdx which is the position index and resourceHash which is hash code of the current node (combination of url, resource name, and description) -->
+  <xsl:template
+    match="*//gmd:MD_DigitalTransferOptions/gmd:onLine
+        [gmd:CI_OnlineResource[gmd:linkage/gmd:URL!=''] and ($resourceIdx = '' or position() = xs:integer($resourceIdx))]
```

Comment:
The position here is computed in the context of `*//gmd:MD_DigitalTransferOptions/gmd:onLine` and during indexing with the context of https://github.com/geonetwork/core-geonetwork/blob/main/schemas/iso19139/src/main/plugin/iso19139/index-fields/index.xsl#L1125-L1126 so the `resourceIdx` is not the same. Removal does not work when we have more than one `transferOptions` block.

Maybe we should simplify the process with the hash only support?
```xsl
  <xsl:template
    match="*//gmd:MD_DigitalTransferOptions/gmd:onLine[digestUtils:md5Hex(normalize-space(.)) = $resourceHash]"
    priority="2">
  </xsl:template>
```

This was also not applied to 115-3 https://github.com/geonetwork/core-geonetwork/blob/main/schemas/iso19115-3.2018/src/main/plugin/iso19115-3.2018/process/onlinesrc-remove.xsl

Reply:
@fxprunayre 
We have recently encountered issues where data was supplied by a third party which has all the resources in separate `transferOptions` block and it was noticed that it was not possible to edit  (and possibly delete) data from these other blocks.  So you are correct that there are issues.

> Maybe we should simplify the process with the hash only support?
- Do you mean ignore the position?  The idea of also checking the position is the following 
  - block1
  - block2
  - block3
  - block4 (Same as block 1 so same hash value)

  Then if someone click `delete` on `block4` then it would delete `block4` even with duplicate hash.
If we only use the hash then it could delete `block1` or `block4`.
I know GN does not allow duplicate data to be entered however when getting data from third party, we don't really know what is being provided.
So position/index was the original goal and the hash was added later as a double check to ensure it did not mess up on the position/index.

  We also have a request from users to be able to order record positions how they like. I think will be easier done if we keep position/index as we should be able to perform a swap based on position? Not really sure as we have not had time to look into this feature yet.

- Do you mean ignore the url check? If so then unchanged metadata 101 schema will stop working if we remove the url check.  This was kept for backward compatibility.

Reply:
> This was also not applied to 115-3
Our intent was to add it but did not get around to it.  Not that there are a couple bugs that we need to fix.  Probably once these are fixed then we can add support to 115-3

Reply:
Yes indeed it depends a bit on how people organize things in ISO distribution blocks or not.

From use cases I have here, I would say hash of (URL+name+protocol) should be unique but user can always create duplicates in different block. We can keep it backward compatible like it is now but if relying on the position, we need to compute a position using the same context (currently the context in indexing is different).

Will think about it, if any better options.